New Rust implementation doesn't accept memoryview

### Things to check first

- [x] I have searched the existing issues and didn't find my bug already reported there

- [x] I have checked that my bug is still present in the latest release


### cbor2 version

6.0.1

### Python version

3.13

### What happened?

The new Rust implementation doesn't accept bytes-like objects. I received this exception via the webauthn package:

```
>           to_return = cbor2.loads(data)
                        ^^^^^^^^^^^^^^^^^
E           TypeError: argument 'data': 'memoryview' object is not an instance of 'bytes'
```

I don't really know much about language bindings - but cbor2 is a python library and as such it would be great if it accepted 'duck-typed' input (which the older implementation did - and in fact explicitly accepted [bytes, bytearray, memoryview]).

Possibly cbor2.loads() should be a python shim that casts/converts python bytes-like input to Rusts strict bytes input.

### How can we reproduce the bug?

```
import cbor2
v = memoryview(b'mycbordata')
cbor2.loads(v)

Traceback (most recent call last):
  File "<input>", line 1, in <module>
TypeError: argument 'data': 'memoryview' object is not an instance of 'bytes'
```

--- a/docs/versionhistory.rst
+++ b/docs/versionhistory.rst
@@ -10,7 +10,10 @@ This library adheres to `Semantic Versioning 2.0 <http://semver.org/>`_.
 - Added the ``allow_duplicate_keys`` parameter to :class:`CBORDecoder`, :func:`load` and
   :func:`loads` (default: ``True``). When set to ``False``, a :exc:`CBORDecodeError` is raised
   upon encountering a duplicate key within the same map.
-  (`#283 <https://github.com/agronholm/cbor2/pull/283>`_)
+  (`#283 <https://github.com/agronholm/cbor2/issues/283>`_)
+- Added support for decoding from any object supporting the buffer API (e.g. ``memoryview`` or
+  ``bytearray``) in addition to ``bytes``
+  (`#297 <https://github.com/agronholm/cbor2/issues/297>`_)
 
 **6.0.1** (2026-04-29)
 

--- a/python/cbor2/__init__.pyi
+++ b/python/cbor2/__init__.pyi
@@ -29,6 +29,11 @@ if sys.version_info >= (3, 13):
 else:
     from typing_extensions import TypeVar
 
+if sys.version_info >= (3, 12):
+    from collections.abc import Buffer
+else:
+    from typing_extensions import Buffer
+
 if sys.version_info >= (3, 11):
     from typing import Self
 else:
@@ -219,7 +224,7 @@ def load(
     immutable: bool = ...,
 ) -> Any: ...
 def loads(
-    data: bytes,
+    data: Buffer,
     *,
     tag_hook: TagHook | None = ...,
     object_hook: ObjectHook | None = ...,

--- a/rust/lib.rs
+++ b/rust/lib.rs
@@ -10,6 +10,8 @@ use pyo3::prelude::pymodule;
 /// A Python module implemented in Rust.
 #[pymodule]
 mod _cbor2 {
+    use pyo3::buffer::PyBuffer;
+    use pyo3::exceptions::PyTypeError;
     use pyo3::prelude::*;
     use pyo3::sync::PyOnceLock;
     use pyo3::types::{PyBytes, PyMapping};
@@ -142,7 +144,10 @@ mod _cbor2 {
     /// Deserialize an object from a bytestring.
     ///
     /// :param data:
-    ///     the bytestring to deserialize
+    ///     the bytestring (or any object implementing the buffer protocol) to deserialize
+    ///
+    ///     .. note:: Types other than :class:`bytes` will be converted to :class:`bytes`, involving
+    ///               memory copying.
     /// :param tag_hook:
     ///     callable that takes 2 arguments: the decoder instance, and the :class:`.CBORTag`
     ///     to be decoded. This callback is invoked for any tags for which there is no
@@ -189,7 +194,7 @@ mod _cbor2 {
     ))]
     fn loads<'py>(
         py: Python<'py>,
-        data: Bound<'py, PyBytes>,
+        data: &Bound<'py, PyAny>,
         tag_hook: Option<&Bound<'py, PyAny>>,
         object_hook: Option<&Bound<'py, PyAny>>,
         semantic_decoders: Option<&Bound<'py, PyMapping>>,
@@ -199,10 +204,23 @@ mod _cbor2 {
         allow_duplicate_keys: bool,
         immutable: bool,
     ) -> PyResult<Bound<'py, PyAny>> {
+        let bytes = if let Ok(bytes) = data.cast::<PyBytes>() {
+            bytes.clone()
+        } else if let Ok(pybuf) = PyBuffer::<u8>::get(data) {
+            PyBytes::new_with(py, pybuf.item_count(), |target| {
+                pybuf.copy_to_slice(py, target)
+            })?
+        } else {
+            return Err(PyTypeError::new_err(format!(
+                "a bytes-like object is required, not '{}'",
+                data.get_type().qualname()?
+            )));
+        };
+
         let mut decoder = CBORDecoder::new_internal(
             py,
             None,
-            Some(data),
+            Some(bytes),
             tag_hook,
             object_hook,
             semantic_decoders,

--- a/tests/test_decoder.py
+++ b/tests/test_decoder.py
@@ -1356,3 +1356,16 @@ def test_override_semantic_decoder() -> None:
 
     payload = unhexlify("c11a514b67b0")  # datetime(2013, 3, 21, 20, 4, 0, tzinfo=timezone.utc)
     assert loads(payload, semantic_decoders={1: date_decoder}) == datetime(2026, 2, 18)
+
+
+@pytest.mark.parametrize(
+    "input_type",
+    [
+        pytest.param(bytes, id="bytes"),
+        pytest.param(bytearray, id="bytearray"),
+        pytest.param(memoryview, id="memoryview"),
+    ],
+)
+def test_loads_buffer_input(input_type: type) -> None:
+    payload = unhexlify("82010a")
+    assert loads(input_type(payload)) == [1, 10]

